ext/liquid_c/tokenizer.c

```diff
@@ -47,28 +48,67 @@ static VALUE tokenizer_initialize_method(VALUE self, VALUE source, VALUE line_nu
     source = rb_str_dup_frozen(source);
     tokenizer->source = source;
     tokenizer->cursor = RSTRING_PTR(source);
-    tokenizer->length = RSTRING_LEN(source);
-    tokenizer->lstrip_flag = 0;
+    tokenizer->cursor_end = tokenizer->cursor + RSTRING_LEN(source);
+    tokenizer->lstrip_flag = false;
     // tokenizer->line_number keeps track of the current line number or it is 0
     // to indicate that line numbers aren't being calculated
-    tokenizer->line_number = RTEST(line_numbers) ? 1 : 0;
+    tokenizer->line_number = FIX2INT(start_line_number);
+    tokenizer->for_liquid_tag = RTEST(for_liquid_tag);
     return Qnil;
 }
 
-void tokenizer_next(tokenizer_t *tokenizer, token_t *token)
+VALUE tokenizer_new_from_cstr(const char *cursor, const char *cursor_end, int line_number, bool for_liquid_tag)
```

Comment:
That's a good catch – it shouldn't be possible at the moment given the lifetime of the temporary tokenizer will be less than or equal to the parent tokenizer, but there's no guarantee that won't change, and it could be intentionally triggered.

Reply:
> it shouldn't be possible at the moment given the lifetime of the temporary tokenizer will be less than or equal to the parent tokenizer

That's not true, since the tokenizer is passed to the tag parse class methods.  Any time an object is passed to ruby code that isn't owned by the library, it is possible that it could hold a reference to that object somewhere and increase its lifetime.  It just isn't likely that the ruby code would do that in this case.

Reply:
I mean, the current Liquid code that exists in Shopify wouldn't have an issue with this.